### DISCUSSION TITLE
How to load external stylesheets from CDN, in SSR using  App router

### DISCUSSION THREAD

**Discussion by ibechev** [vercel/next.js#51230] [Help] [answered]
### Summary

I'm exploring a possible migration to NextJs. The application needs to support themes, which should be configurable at run time, meaning that the theme should be loaded and applied on every SSR. 

The app uses `css-modules`. I'm looking for a way to inject global css variables in the HTML (head) that comes from the server. The values of the variables should be loaded on every app reload, as opposed to being in a static file. The goal is to use these variables as a theme provider across the application.  

I tried to use `@import url(https:// ...)` directly in `globals.css` but it doesn't seem to work. However if I place it in any other `.module.css` file it does get loaded. 

I know that `<link>` and `<style>` tags are not allowed the `<Head>` from next. 

Any suggestions? 

### Additional information

```js
next: 13.4.2
```

### Example
Example of the result I'm trying to achieve:

`/src/app/layout.tsx`
```
import './globals.css';
import Head from 'next/head'

interface LayoutProps {
    children: React.ReactNode;
}

const loadThemeFromExternalSource = () => ({
    primary: '#4d216d',
    secondary: '#eef2f5',
});

const Layout = ({ children }: LayoutProps) => {
    const theme = loadThemeFromExternalSource();

    return (
        <html lang="en">
            <Head>
                <script>{`
                :root {
                    --color-primary: ${theme.primary};
                    --color-secondary: ${theme.secondary};
                }
                `}</script>
            </Head>
            <body>
                <header>Header</header>
                {children}
                <footer>Footer</footer>
            </body>
        </html>
    );
};

export default Layout;

```

**Comment by devjiwonchoi** [comment, marked as answer]
> I know that <link> and <style> tags are not allowed the <Head> from next.

They are allowed in the `<head>` tag.

For `link` tags that are not provided by default in [Metadata](https://nextjs.org/docs/app/building-your-application/optimizing/metadata), you can simply add them inside the `<head>` tag above the body.

```js
import './globals.css'

interface LayoutProps {
  children: React.ReactNode
}

const loadThemeFromExternalSource = () => ({
  primary: '#4d216d',
  secondary: '#eef2f5',
})

const Layout = ({ children }: LayoutProps) => {
  const theme = loadThemeFromExternalSource()

  return (
    <html lang='en'>
      <head>
        <style>
          {`
            :root {
                --color-primary: ${theme.primary};
                --color-secondary: ${theme.secondary};
            }
          `}
        </style>
        <link
          rel='stylesheet'
          href='https://cdn.jsdelivr.net/gh/devjiwonchoi/simple-kakaomap@main/simple-kakaomap.css'
        />
      </head>
      <body>
        <header>Header</header>
        <div className='map_wrap'>hi</div>
        {children}
        <footer>Footer</footer>
      </body>
    </html>
  )
}

export default Layout

```

Although the docs state that the external stylesheet using `<link rel='stylesheet' />` is not allowed, it seems like to successfully imported. Please let me know if I've been mistaken. (Maybe it's because it's not a `.module.css`?)

> External stylesheets must be directly imported from an npm package or downloaded and colocated with your codebase. You cannot use <link rel="stylesheet" />.

Your `theme` modification seems to be working fine,

![Screenshot 2023-06-13 at 9 02 49 PM](https://github.com/vercel/next.js/assets/120007119/a83eacc2-22a7-4297-ab7c-9ab928b72c78)

and my external cdn css seems to be working fine also.

![Screenshot 2023-06-13 at 9 03 50 PM](https://github.com/vercel/next.js/assets/120007119/6374b5fb-0e4e-4f19-8ebe-9fe3db78c2c2)

## UPDATE
This answer is invalid. Please see my new answer below. 
https://github.com/vercel/next.js/discussions/51230#discussioncomment-6164648

## Reference
- [Metadata](https://nextjs.org/docs/app/building-your-application/optimizing/metadata)
- [External Stylesheets](https://nextjs.org/docs/app/building-your-application/styling/css-modules#external-stylesheets)

**Comment by ibechev** [comment]
@devjiwonchoi thank you for the reply. 

In relation to the `<script>` tag, yes you are right, it does work and it adds declared variables globally. However, there is a hydration error for unmatched server and client text, which leads to the whole HTML being re-rendered on the client: 

![image](https://github.com/vercel/next.js/assets/2820264/925f0312-4cda-4ced-a556-20af31ccf30a)

In relation of the `<link>` tag in the `<head>` it does work and there are no errors thrown. The only concern I have  is that in [the docs](https://nextjs.org/docs/app/building-your-application/styling/css-modules#external-stylesheets) its written that the `<link>` tag can not be used. Even thou it works, I'm not fully confident in using it.

**Reply by devjiwonchoi** [reply]
You need to manipulate the ReactDom to safely insert it into the `head` tag.

> you can use new [ReactDOM methods](https://github.com/facebook/react/pull/26237) to safely insert them into the <head> of the document.

```js
// app/preload-resources.ts
'use client'

import ReactDOM from 'react-dom'

export function PreloadResources() {
  ReactDOM.preload('CDN_CSS_HREF', { as: 'style' })

  return null
}
```

Note that the option `as` needs to be one of these: `'font' | 'script' | 'style'`

## Reference
- [Resource hints](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#resource-hints)

**Reply by ibechev** [reply]
Thank you @devjiwonchoi, at the end I decided to go with placing a `<link>` inside a `<head>` tag. I've also tried using `preload-resources` but for some reason it didn't work.
